### PR TITLE
feat: smart contract revision

### PR DESCRIPTION
**Pull request by jorgeantonio21**
1. Makes tasks permissioned;
2. Hardcodes max compute units a node can subscribe to (currently 2_560_000);
3. Refactored methods `subscribe_node_to_task` and `update_node_subscription` accordingly (in both contract and cli);
4. Rename price per compute unit to price per one million compute units (which works well for both chat completions, text embeddings and image generation)

### CODE REVIEW

File: sui/packages/atoma/sources/db.move [hunk 1]
```diff
@@ -418,7 +420,7 @@ module atoma::db {
         deprecated_at_epoch: Option<u64>,
         /// Security level for the task
         security_level: SecurityLevel,
-        /// Subscribed nodes table, where key is node SmallId and value is price per compute unit
+        /// Subscribed nodes table, where key is node SmallId and value is price per one million comupte units
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:423]:
```suggestion
        /// Subscribed nodes table, where key is node SmallId and value is price per one million compute units
```

File: sui/packages/atoma/sources/db.move [hunk 2]
```diff
@@ -228,8 +230,8 @@ module atoma::db {
         selected_node_id: NodeSmallId,
         /// The number of compute units allocated to this stack
         num_compute_units: u64,
-        /// The price per compute unit in TOMA tokens
-        price: u64,
+        /// The price per one million comupte units in TOMA tokens
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:233]:
```suggestion
        /// The price per one million compute units in TOMA tokens
```

File: sui/packages/atoma/sources/db.move [hunk 3]
```diff
@@ -449,11 +451,11 @@ module atoma::db {
         inner: u8,
     }
 
-    /// Data about a node's price per compute unit for a task
+    /// Data about a node's price per one million comupte units for a task
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:454]:
```suggestion
    /// Data about a node's price per one million compute units for a task
```

File: sui/packages/atoma/sources/db.move [hunk 4]
```diff
@@ -449,11 +451,11 @@ module atoma::db {
...
     public struct NodePriceData has store, copy, drop {
         node_id: NodeSmallId,
-        /// Price per compute unit in TOMA for the current task
-        price_per_compute_unit: u64,
+        /// price per one million comupte units in TOMA for the current task
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:457]:
```suggestion
        /// price per one million compute units in TOMA for the current task
```

File: sui/packages/atoma/sources/db.move [hunk 5]
```diff
@@ -1203,37 +1208,37 @@ module atoma::db {
...
-            max_num_compute_units,
+            price_per_one_million_compute_units,
+            max_num_compute_units: COMPUTE_UNITS_PER_STACK,
         });
     }
 
     /// Updates a node's subscription details for a specific task.
     ///
-    /// This function allows a node to modify its price per compute unit and maximum number of compute units
+    /// This function allows a node to modify its price per one million comupte units and maximum number of compute units
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1218]:
```suggestion
    /// This function allows a node to modify its price per one million compute units and maximum number of compute units
```

File: sui/packages/atoma/sources/db.move [hunk 6]
```diff
@@ -1203,37 +1208,37 @@ module atoma::db {
...
     /// for a task it's already subscribed to. This enables nodes to dynamically adjust their pricing and
     /// capacity based on current conditions or strategy.
     ///
     /// # Arguments
     /// * `self` - A mutable reference to the AtomaDb object.
     /// * `node_badge` - A mutable reference to the NodeBadge of the node updating its subscription.
     /// * `task_small_id` - The SmallId of the task for which the subscription is being updated.
-    /// * `price_per_compute_unit` - The new price per compute unit in TOMA tokens.
+    /// * `price_per_one_million_compute_units` - The new price per one million comupte units in TOMA tokens.
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1226]:
```suggestion
    /// * `price_per_one_million_compute_units` - The new price per one million compute units in TOMA tokens.
```

File: sui/packages/atoma/sources/db.move [hunk 7]
```diff
@@ -1203,37 +1208,37 @@ module atoma::db {
...
     /// * `max_num_compute_units` - The new maximum number of compute units the node is willing to process.
     ///
     /// # Errors
     /// * `ETaskNotFound` - If the specified task does not exist in the AtomaDb.
     /// * `ENodeNotSubscribedToTask` - If the node is not currently subscribed to the specified task.
     ///
     /// # Effects
-    /// - Updates the node's price per compute unit and maximum number of compute units for the specified task.
+    /// - Updates the node's price per one million comupte units and maximum number of compute units for the specified task.
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1234]:
```suggestion
    /// - Updates the node's price per one million compute units and maximum number of compute units for the specified task.
```

File: sui/packages/atoma/sources/db.move [hunk 8]
```diff
@@ -1203,37 +1208,37 @@ module atoma::db {
...
     /// - Emits a `NodeSubscriptionUpdatedEvent` with the updated subscription details.
     ///
     /// # Events
     /// Emits a `NodeSubscriptionUpdatedEvent` containing:
     /// - `node_small_id`: The SmallId of the node updating its subscription.
     /// - `task_small_id`: The SmallId of the task being updated.
-    /// - `price_per_compute_unit`: The new price per compute unit.
+    /// - `price_per_one_million_compute_units`: The new price per one million comupte units.
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1241]:
```suggestion
    /// - `price_per_one_million_compute_units`: The new price per one million compute units.
```

File: sui/packages/atoma/sources/db.move [hunk 9]
```diff
@@ -1340,7 +1344,7 @@ module atoma::db {
     /// * `wallet` - A mutable reference to the Coin<TOMA> object to pay for the stack.
     /// * `task_small_id` - The SmallId of the task associated with this stack.
     /// * `num_compute_units` - The number of compute units allocated to this stack.
-    /// * `price` - The price per compute unit in TOMA tokens.
+    /// * `price` - The price per one million comupte units in TOMA tokens.
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1347]:
```suggestion
    /// * `price` - The price per one million compute units in TOMA tokens.
```

File: sui/packages/atoma/sources/db.move [hunk 10]
```diff
@@ -1429,9 +1433,9 @@ module atoma::db {
             let fee_amount = if (security_level.inner == SamplingConsensus) {
                 let sampling_consensus_charge = self.get_sampling_consensus_charge_permille();
                 let cross_validation_charge = self.get_cross_validation_extra_nodes_charge_permille();
-                (price * num_compute_units * (sampling_consensus_charge + cross_validation_charge)) / 1000
+                (price_per_one_million_compute_units * num_compute_units * (sampling_consensus_charge + cross_validation_charge)) / 1000
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1436]: Shouldn't this fee reflect that the price is for 1 mil ?

File: sui/packages/atoma/sources/db.move [hunk 11]
```diff
@@ -1429,9 +1433,9 @@ module atoma::db {
...
             } else {
-                (price * num_compute_units)
+                (price_per_one_million_compute_units * num_compute_units)
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:1438]: same here

File: sui/packages/atoma/sources/db.move [hunk 12]
```diff
@@ -2566,7 +2598,7 @@ module atoma::db {
     ///
     /// # Arguments
     /// * `security_level` - The security level of the task (e.g., ascii::string(b"sampling-consensus")).
-    /// * `stack_price` - The price per compute unit for the stack.
+    /// * `stack_price` - The price per one million comupte units for the stack.
```
Comment (Cifko) [inline sui/packages/atoma/sources/db.move:2601]:
```suggestion
    /// * `stack_price` - The price per one million compute units for the stack.
```

File: sui/packages/atoma/sources/db.move (hunk 1 above)
Comment (jorgeantonio21) [reply sui/packages/atoma/sources/db.move:423]: Oh nice catch !

File: sui/packages/atoma/sources/db.move (hunk 10 above)
Comment (jorgeantonio21) [reply sui/packages/atoma/sources/db.move:1436]: oh yeah ! Great catch !

**Review by Cifko — APPROVED** [review]

**Pull request merged**